main.tf

```diff
@@ -18,11 +18,12 @@
   Bucket random id suffix configuration
  *****************************************/
 resource "random_id" "bucket_suffix" {
+  count       = var.randomize_suffix ? 1 : 0
   byte_length = 2
 }
 
 locals {
-  suffix       = var.randomize_suffix ? random_id.bucket_suffix.hex : ""
+  suffix       = var.randomize_suffix ? random_id.bucket_suffix[0].hex : ""
```

Comment:
FYI: This will be a breaking change for existing users.  Probably can use something like `terraform state mv 'random_id.bucket_suffix.hex' 'random_id.bucket_suffix[0].hex'` to migrate the existing suffix to the new resource location.

Reply:
It won't, Terraform implicitly modifies the state. Try to run the following experiment:

1. Draft a simple configuration without `count` and apply it:

    ```hcl
    resource "random_id" "x" {
      byte_length = 1
    }
    ```
    ```shell
    $ terraform apply -auto-approve
    Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
      + create
    
    Terraform will perform the following actions:
    
      # random_id.x will be created
      + resource "random_id" "x" {
          + b64_std     = (known after apply)
          + b64_url     = (known after apply)
          + byte_length = 1
          + dec         = (known after apply)
          + hex         = (known after apply)
          + id          = (known after apply)
        }
    
    Plan: 1 to add, 0 to change, 0 to destroy.
    random_id.x: Creating...
    random_id.x: Creation complete after 0s [id=wA]
    
    Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
    ```

1. Modify resource with `count` and apply it:

    ```hcl
    resource "random_id" "x" {
     count = 1
     byte_length = 1
    }
    ```
    ```shell
    $ terraform apply -auto-approve
    random_id.x[0]: Refreshing state... [id=wA]
    
    Terraform will perform the following actions:
    
      # random_id.x has moved to random_id.x[0]
        resource "random_id" "x" {
            id          = "wA"
            # (5 unchanged attributes hidden)
        }
    
    Plan: 0 to add, 0 to change, 0 to destroy.
    
    Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
    ```

Pay attention resource has been moved, not replacement happened.

Reply:
Thanks for explanation @Tensho!

I just confirmed this worked on TF 0.13 (https://github.com/terraform-google-modules/terraform-google-cloud-storage/blob/master/versions.tf#L18), although the messaging is not as clear:

```
./terraform apply
random_id.x[0]: Refreshing state... [id=9w]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```